errors.go

```diff
@@ -286,3 +286,13 @@ func Cause(err error) error {
 	}
 	return err
 }
+
+func Unwrap(err error) error {
```

Comment:
Why reimplement this function rather than wrapping it? Someone using Go 1.12 or older will not be using the new style of errors. This creates a risk of incompatibility if Go 1.14 subtly changes its `Unwrap` implementation.

Reply:
@aperezg emm.... this is a problem if go1.14 changes `Unwrap` implment, so... we should keep `Unwrap` >= go1.13?

Reply:
But why is the problem to maintain the Unwrap function like today? I understand that the Unwrap is an interface on a std library right? Sorry, I try to understand what is the point to change the implementation a wrapped with the standard library. I doubt that they implement a breaking change on that, no?

https://github.com/pkg/errors/blob/master/errors.go#L163

Reply:
go1.14 cannot break the semantics of `Unwrap` without violating the go1 compat guarantee.

> These same considerations apply to successive point releases. For instance, code that runs under Go 1.2 should be compatible with Go 1.2.1, Go 1.3, Go 1.4, etc., although not necessarily with Go 1.1 since it may use features added only in Go 1.2

Reply:
This functionality should likely be folded into `Cause()` and `Unwrap()` here just calls `Cause()`.

c.f. https://github.com/pkg/errors/pull/215

Reply:
> go1.14 cannot break the semantics of Unwrap without violating the go1 compat guarantee.

Right now, standard library's `errors.Unwrap` is only interested in an `Unwrap()` interface, and it will always have to do that to retain backwards compatibility. However, Go 1.14 could hypothetically update `errors.Unwrap` to *additionally* handle a new `Foobar()` interface. In that scenario, it would be painful for `pkg/errors` to roll its own implementation for a "drop-in replacement".

Reply:
I think worrying too much into the future results in writing features and support for things that maybe never even are necessary.

We should provide a `Cause`/`Unwrap` in this library that supports _both_ this package’s wrapping, as well as the standard package wrapping. We should concern ourselves with _today_.

If go1.14 expands the functionality of `Wrap`, then we can deal with that change _then_.

It does not make sense to burn bridges before we even get to them.

Reply:
Hypothetical edge cases aside, calling through to the standard library's `Unwrap` will support both flavors of wrapped errors, and it's simpler than reimplementing the function. I think the latest commit in this PR is doing the right thing in that regard.

Reply:
in my option, >1.13 errors will be compatible with xerrors, and xerrors now still uses unwrap interface. if std errors unwrap change its implementation, it will break old xerrors user codes

Reply:
and `withStack` `withMessage` implement the unwrap interface to support these functions

Reply:
Right, `xerrors.Unwrap()` will support this package as well, but what about a package that only supports `causer`?

Reply:
that is bad idea, adding these functions is trying to make sure we don't import two packages at the same time, it may confuse user when use pkg/errors when std errors

Reply:
It does not matter what _we_ want, this “theoretical” is a third-party package that only implements `causer`, has been abandoned and no longer updated. So it will never implement “unwrapper”.

So, again. It does not matter what _we_ implement in this package, for `withMessage` and `withStack`. If some other third-party package implemented `causer` but not `unwrapper`… we still want to ourselves allow some way to extract the original error from both `causer`s and `unwrapper`s in any permutation.

Reply:
```go
func Unwrap(err) error {
    if unwrapped := errors.Unwrap(err); unwrapped != nil {
        return unwrapped
    }
    if causer, ok := err.(causer); ok {
        return causer.Cause()
    }
    return nil
}
```

Reply:
@jayschwa that is great, if we unwrap an error from other lib and it uses old pkg errors, the unwrap can still return the original error

Reply:
as [that](https://github.com/pkg/errors/pull/213#discussion_r345502202) says, 1.14 errors won't change implementation, so we could make unwrap continue in normal package @aperezg

Reply:
That `Unwrap` only works for specific permutations of wrapping with the standard library and `causer`s.

In the wild, we should expect any possible permutation of wrapping, including multiple times being wrapped in an alternating manner.

Reply:
`Unwrap` should only unwrap error one times, no matter the inner error implement unwrap interface or not, like std `Unwrap` do.

Reply:
But this code would fail to fully unwrap: `Unwrap(fmt.Errorf("wrap1: %w", myPrivateCauserOnlyError(fmt.Errorf("wrap 3: %w", err), "wrap2:")))` where I would get the `fmt.Errorf("wrap3: %w", err)` error not the original `err`, which would not be the expected behavior.

Reply:
🤦‍♀ the standard library `Unwrap` does not continuously unwrap a function, until it reaches the end result, unlike how `Cause` in this package works.